### DISCUSSION TITLE
Structure for a block created specifically for controlling the lifetime of RAII object.

### DISCUSSION THREAD

**Discussion by gregmarr** [hsutter/cppfront#682] [Suggestions]
This is based on a recent real-life internal discussion.

Problem statement: user wants to make it more explicit why a particular scope block exists in the code.

Current methods:

```
    int foo()
    {
        do_something();

        { // scope to control lifetime of scoped_lock
            auto sl = scoped_lock(blah);
            do_stuff();
        }

        do_something_else();
    }
``` 

```
    int foo()
    {
        do_something();

        do { // scope to control lifetime of scoped_lock
            auto sl = scoped_lock(blah);
            do_stuff();
        } while(false);

        do_something_else();
    }
``` 

Newer method that someone thought about but didn't really like:

```
    int foo()
    {
        do_something();

        if (auto sl = scoped_lock(blah); true) // linters may complain about this because of the constant condition
        {
            do_stuff();
        }

        do_something_else();
    }
``` 

Someone suggested that it would be nice to have something similar to Python's "with"

```
    int foo()
    {
        do_something();

        with (auto sl = scoped_lock(blah))
        {
            do_stuff();
        }

        do_something_else();
    }
``` 

or if the RAII object doesn't need to be accessed, and just needs a lifetime scope:

```
    int foo()
    {
        do_something();

        with (scoped_lock(blah))
        {
            do_stuff();
        }

        do_something_else();
    }
``` 

There were a couple attempts to emulate this with lambdas and helper structs, but they weren't pretty.  I'm interested to see what kinds of solutions might already exist with Cpp2, or if it's something that could be added.

**Comment by JohelEGP** [comment]
In Cpp2, you can use a [parameterized statement](https://github.com/hsutter/cppfront/issues/386#issuecomment-1518194517) (<https://cpp2.godbolt.org/z/16rz7YqPr>):
```Cpp2
#include <mutex>
main: () = {
  m := std::mutex();
  (_ := std::scoped_lock(m)) {
    // ...
  }
}
```

**Reply by gregmarr** [reply]
Thanks, I had a feeling that there was something that covered this, but I couldn't come up with it.  I'm remembering the discussion now.
